scheduler/generic_sched.go

```diff
@@ -529,10 +529,6 @@ func (s *GenericScheduler) computePlacements(destructive, place []placementResul
 				// If we are placing a canary and we found a match, add the canary
 				// to the deployment state object and mark it as a canary.
 				if missing.Canary() && s.deployment != nil {
-					if state, ok := s.deployment.TaskGroups[tg.Name]; ok {
-						state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
-					}
```

Comment:
I'm unclear why an inplace update is inappropriate or the source of a bug here. It's hard to follow this flow, but `s.deployment` is set by the reconciler which creates a *copy* of the deployment:

https://github.com/hashicorp/nomad/blob/f5441e69d4c991b77318bb661b31ab6acb2c7ea7/scheduler/reconcile.go#L170

This seems like an intentional decision to make deployments safe for inplace updates, and it appears `s.deployment == s.Plan.Deployment` which means it will get submitted along with the plan.

If that's not the case I'd love to see a little more cleanup or commenting in the code to help future developers. Maybe an explicit comment on the `s.deployment` field about its mutability, and removing the `deployment.Copy()` if it is immutable.

Reply:
So there is definitely some misdirection going on with all these `deployment`s I'm going to outline my understanding of them and then see about commenting or cleaning up.

>  but s.deployment is set by the reconciler which creates a copy of the deployment

The scheduler deployment is actually not set by the reconciler, the reconciler copies the scheduler deployment to its own `deployment` and uses it as a readonly reference to the current state of the deployment and returns a `results.deployment` and `results.deploymentUpdates`

https://github.com/hashicorp/nomad/blob/master/scheduler/generic_sched.go#L355-L357

Currently the reconciler will only return a `results.deployment` if the deployment is newly created, otherwise it will just be `deploymentUpdates`.

So the following logic is mutating state that never gets written to the raft log because only changes on `s.plan` are ultimately submitted, in a single node cluster the state will remain locally modified and show the correct number of canaries but in multi node clusters a different worker won't have these changes
```go
if state, ok := s.deployment.TaskGroups[tg.Name]; ok {
	state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
}
```